zebra-state/src/lib.rs

```diff
@@ -27,20 +27,58 @@ pub mod on_disk;
 #[serde(deny_unknown_fields)]
 pub struct Config {
     /// The root directory for the state storage
-    pub path: PathBuf,
+    pub path: Option<PathBuf>,
 }
 
 impl Config {
+    /// Generate the appropriate `sled::Config` based on the provided
+    /// `zebra_state::Config`.
+    ///
+    /// # Details
+    ///
+    /// This function should panic if the user of `zebra-state` doesn't configure
+    /// a directory to store the state.
+    ///
+    /// Currently this is only partially implemented by leveraging a `SpanTrace`
+    /// to print a TODO to the user which isn't as nice as what we'd eventually
+    /// want but it definitely gets the point across.
+    ///
+    /// # Current Format
+    ///
+    /// <pre><font color="#CC0000">The application panicked (crashed).</font>
+    /// Message:  <font color="#06989A">called `Option::unwrap()` on a `None` value</font>
+    /// Location: <font color="#75507B">zebra-state/src/lib.rs</font>:<font color="#75507B">40</font>
+    ///
+    /// Backtrace omitted.
+    ///
+    /// Run with RUST_BACKTRACE=1 environment variable to display it.
+    /// Run with RUST_BACKTRACE=full to include source snippets.
+    ///   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ SPANTRACE ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
+    ///
+    ///    0: <font color="#F15D22">zebra_state::sled_config::comment</font> with <font color="#34E2E2">text=TODO(jlusby): Replace this unwrap with a nice user-facing</font>
+    /// <font color="#34E2E2">   explaining error that the cache dir must be specified</font>
+    ///       at <font color="#75507B">zebra-state/src/lib.rs</font>:<font color="#75507B">36</font></pre>
+    #[spandoc::spandoc]
     pub(crate) fn sled_config(&self) -> sled::Config {
-        sled::Config::default().path(&self.path)
+        /**
+         * SPANDOC: TODO(jlusby): Replace this unwrap with a nice user-facing
+         * explaining error that the cache dir must be specified
+         */
+        let path = self.path.as_ref().unwrap().join("state");
+
+        sled::Config::default().path(path)
     }
 }
 
 impl Default for Config {
     fn default() -> Self {
-        Self {
-            path: PathBuf::from("./.zebra-state"),
-        }
+        let path = std::env::var("ZEBRAD_CACHE_DIR")
+            .map(PathBuf::from)
+            .ok()
+            .or_else(dirs::cache_dir)
+            .map(|dir| dir.join("zebra"));
```

Comment:
@teor2345  I configured this so that we default to `$CACHE_DIR/zebra` which can be overridden and I moved the `state` join to the `sled_config` fn so that no matter what dir you specify as the zebra cache dir the state will always be isolated to a dir called "state" within that directory.

I am a little on the fence if I should always add the `zebra` onto the path. When we're basing the config out of the `cache_dir` I think we would definitely want to append a `zebra` but you could argue that `ZEBRAD_CACHE_DIR` should be equivalent to `dirs::cache_dir()/zebra` rather than being equivalent to `dirs::cache_dir()` itself. I considered just reading `CACHE_DIR` instead of `ZEBRAD_CACHE_DIR` but I'm worried about that env variable being set while not intended for `zebrad`. LMK what you think

Reply:
I think we should add `zebra` in generic cache directories, but not for `ZEBRAD_CACHE_DIR`.

(And we should document that as part of documenting ZEBRAD_CACHE_DIR, so it's less surprising.)